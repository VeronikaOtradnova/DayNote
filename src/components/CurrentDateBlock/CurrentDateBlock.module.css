.wrapper {
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  padding: 10px;
}

.date {
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
  align-items: center;
  width: 50%;
  height: 45px;
  padding: 0 10px;
  border-radius: 13px;
  background-color: #fff;
  font-size: 2rem;
  color: #000;
}
.date:hover {
  cursor: pointer;
  box-shadow: 0px 0px 8px 0px #cccccc;
}

.calendarIcon {
  width: 22px;
  height: 24px;
  background-color: transparent;
  background-image: url('/src/icons/calendar.svg');
  background-size: 100% 100%;
}

.dateText {
  margin: auto;
}

.btn {
  width: 45px;
  height: 45px;
  border-radius: 13px;
  background-color: #fff;
  background-image: url('/src/icons/chevron.svg');
  background-size: 18px 18px;
  background-position: center center;
  background-repeat: no-repeat;
  border: none;
  outline: none;
}
.btn:hover {
  cursor: pointer;
  box-shadow: 0px 0px 8px 0px #cccccc;
}

.prevDateBtn {
  transform: rotate(-90deg);
}

.nextDateBtn {
  transform: rotate(90deg);
}

@media (min-width: 450px) {
  .date {
    width: auto;
    height: 65px;
    font-size: 2.4rem;
  }

  .calendarIcon {
    width: 35px;
    height: 37px;
  }

  .dateText {
    margin: 0 55px;
  } 

  .btn {
    width: 65px;
    height: 65px;
  }
}