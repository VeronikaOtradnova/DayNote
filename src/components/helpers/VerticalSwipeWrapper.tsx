import { useState } from "react"

interface IProps {
  children: React.ReactNode,
  onTop?: () => void,
  onBottom?: () => void,
}

export function VerticalSwipeWrapper({children, onTop = () => {}, onBottom = () => {}}: IProps) {
  const [touchStart, setTouchStart] = useState<number | null>(null)
  const [touchEnd, setTouchEnd] = useState<number | null>(null)

  // the required distance between touchStart and touchEnd to be detected as a swipe
  const minSwipeDistance = 50

  const onTouchStart = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(null)
    setTouchStart(e.targetTouches[0].clientY)
  }

  const onTouchMove = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(e.targetTouches[0].clientY);
  };

  const onTouchEnd = () => {
    if (!touchStart || !touchEnd) return
    const distance = touchStart - touchEnd
    const isTopSwipe = distance > minSwipeDistance
    const isBottomSwipe = distance < -minSwipeDistance
    if (isTopSwipe || isBottomSwipe) {
      isTopSwipe ? onTop() : onBottom()
    }
  }

  return (
    <div 
      onTouchStart={onTouchStart} 
      onTouchMove={onTouchMove} 
      onTouchEnd={onTouchEnd} 
    >
      {children}
    </div>
  )
}