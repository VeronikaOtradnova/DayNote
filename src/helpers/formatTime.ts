export function formatTime(timeInMillis: number) {
  const date = new Date(timeInMillis);
  
  // Получаем часы и минуты из объекта Date
  let hours = date.getUTCHours().toString();
  let minutes = date.getUTCMinutes().toString();
  
  // Форматируем часы и минуты в виде двухзначных чисел
  hours = hours.padStart(2, '0');
  minutes = minutes.padStart(2, '0');
  
  // Возвращаем строку в формате 'hh:mm'
  return `${hours}:${minutes}`;
}