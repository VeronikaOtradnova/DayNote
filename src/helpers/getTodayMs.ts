export const getTodayMs = () => {
  return +(new Date().setHours(0, 0, 0, 0));
}