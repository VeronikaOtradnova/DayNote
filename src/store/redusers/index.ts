import { combineReducers } from "redux";
import { dayReducer } from "./dayReducer";
import { calendarReducer } from "./calendarReducer";

export const rootReducer = combineReducers({
  day: dayReducer,
  calendar: calendarReducer
})

export type TRootState = ReturnType<typeof rootReducer>