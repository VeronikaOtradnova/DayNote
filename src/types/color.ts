export enum colors {
  none = 'none',
  green = 'green',
  yellow = 'yellow',
  gray = 'gray',
}

export type TColor = colors.green | colors.yellow | colors.gray | colors.none;