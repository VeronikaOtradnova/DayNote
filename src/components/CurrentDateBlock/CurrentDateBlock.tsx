import { getDayMs } from '../../helpers/getDayMs';
import { getTodayMs } from '../../helpers/getTodayMs';
import { useActions } from '../../hooks/useActions';
import { useCurrentDayData } from '../../hooks/useCurrentDayData';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import styles from './CurrentDateBlock.module.css';

export const CurrentDateBlock = () => {
  const { currentDay } = useTypedSelector(store => store.day);
  const {setCurrentDay, openCalendar} = useActions();
  const {color} = useCurrentDayData();

  if (!currentDay) return null;

  const day = new Date(currentDay).getDate();
  const numberMonth = new Date(currentDay).getMonth();
  let month;
  switch (numberMonth) {
    case 0:
      month = 'января';
      break;
    case 1:
      month = 'февраля';
      break;
    case 2:
      month = 'марта';
      break;
    case 3:
      month = 'апреля';
      break;
    case 4:
      month = 'мая';
      break;
    case 5:
      month = 'июня';
      break;
    case 6:
      month = 'июля';
      break;
    case 7:
      month = 'августа';
      break;
    case 8:
      month = 'сентября';
      break;
    case 9:
      month = 'октября';
      break;
    case 10:
      month = 'ноября';
      break;
    case 11:
      month = 'декабря';
      break;
  }

  const prevBtnHandler = () => {
    setCurrentDay(currentDay - getDayMs());
  }

  const nextBtnHandler = () => {
    setCurrentDay(currentDay + getDayMs());
  }

  return (
    <div className={styles.wrapper} data-testid="date-block">
      <button 
        onClick={prevBtnHandler} 
        className={`${styles.btn} ${styles.prevDateBtn} ${color ? styles[color] : styles.noColor}`} 
        data-testid="prev-btn" 
      />

      <div className={`${styles.date} ${color ? styles[color] : styles.noColor}`} onClick={() => openCalendar()} data-testid="date-elem">
        <div className={styles.calendarIcon} />
        <div className={styles.dateText}>{`${day} ${month}`}</div>
      </div>
      {
        currentDay === getTodayMs() ?
        <button 
          disabled 
          className={`${styles.btn} ${styles.nextDateBtn} ${color ? styles[color] : styles.noColor} disabled-btn`} 
          data-testid="disabled-next-btn" 
        /> 
        :
        <button 
          onClick={nextBtnHandler} 
          className={`${styles.btn} ${styles.nextDateBtn} ${color ? styles[color] : styles.noColor}`} 
          data-testid="next-btn" 
        />
      }
    </div>
  )
}