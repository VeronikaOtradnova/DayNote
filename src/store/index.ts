import { applyMiddleware, createStore } from "redux";
import { rootReducer } from "./redusers";
import { thunk } from "redux-thunk";

const stateFromStorage = localStorage.getItem('reduxState')
export const persistedState = stateFromStorage ? JSON.parse(stateFromStorage): {}

export const store = createStore(rootReducer, persistedState, applyMiddleware(thunk));

export type AppDispatch = typeof store.dispatch;

store.subscribe(()=>{
  localStorage.setItem('reduxState', JSON.stringify(store.getState()))
})