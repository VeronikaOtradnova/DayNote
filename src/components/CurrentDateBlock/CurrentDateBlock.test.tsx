import { screen } from "@testing-library/react"
import { renderWithRedux } from "../../tests/helpers/renderWithRedux"
import { colors } from "../../types/color"
import { CurrentDateBlock } from './CurrentDateBlock'
import { TRootState } from "../../store/redusers"
import userEvent from "@testing-library/user-event"
import { getTodayMs } from "../../helpers/getTodayMs"
import { act } from "react-dom/test-utils"

const testInitialState: TRootState = {
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
      tasks: []
    }]
  },
  calendar: {
    isCalendarOpen: false,
    calendarDate: +(new Date(2020, 0, 1)),
  }
}

const testInitialStateToday: TRootState = {
  day: {
    currentDay: getTodayMs(),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
      tasks: []
    }]
  },
  calendar: {
    isCalendarOpen: false,
    calendarDate: +(new Date(2020, 0, 1)),
  }
}

describe('CURRENT-DATE-BLOCK TESTING', () => {
  test('Render current date', () => {
    renderWithRedux(<CurrentDateBlock />, { initialState: testInitialState });

    const dateBlock = screen.getByTestId('date-block');
    expect(dateBlock).toBeInTheDocument();
    expect(dateBlock).toMatchSnapshot();

    expect(screen.getByTestId<HTMLDivElement>('date-elem')).toHaveTextContent('1 января');
  })

  test('Prev-date btn', async () => {
    const user = userEvent.setup();
    renderWithRedux(<CurrentDateBlock />, { initialState: testInitialState });

    const prevBtn = screen.getByTestId<HTMLButtonElement>('prev-btn');
    expect(prevBtn).toBeInTheDocument();

    await act(async () => {
      await user.click(prevBtn);
    })

    expect(screen.getByTestId<HTMLDivElement>('date-elem')).toHaveTextContent('31 декабря');
  });

  test('Next-date btn', async () => {
    const user = userEvent.setup();
    renderWithRedux(<CurrentDateBlock />, { initialState: testInitialState });

    const nextBtn = screen.getByTestId<HTMLButtonElement>('next-btn');
    expect(nextBtn).toBeInTheDocument();

    await act(async () => {
      await user.click(nextBtn);
    })
    
    expect(screen.getByTestId<HTMLDivElement>('date-elem')).toHaveTextContent('2 января');
  });

  test('Disabled next-date btn', async () => {
    const user = userEvent.setup();
    const {store} = renderWithRedux(<CurrentDateBlock />, { initialState: testInitialStateToday });
    const firstState:TRootState = store.getState();

    expect(screen.queryByTestId('next-btn')).not.toBeInTheDocument();
    const disabledBtn = screen.getByTestId<HTMLButtonElement>('disabled-next-btn');
    expect(disabledBtn).toBeInTheDocument();

    await act(async () => {
      await user.click(disabledBtn);
    })
    
    const stateAfterClick:TRootState = store.getState();
    expect(firstState.day.currentDay).toBe(stateAfterClick.day.currentDay); // после клика дата в стейте не изменилась
  });

  test('Should change state.calendar.icCalendarOpen to true after click on date-elem', async () => {
    const user = userEvent.setup();
    const {store} = renderWithRedux(<CurrentDateBlock />, { initialState: testInitialState });

    const dateElem = screen.getByTestId('date-elem');
    
    await act(async () => {
      await user.click(dateElem);
    })

    const state:TRootState = store.getState();
    expect(state.calendar.isCalendarOpen).toBe(true);
  })
})