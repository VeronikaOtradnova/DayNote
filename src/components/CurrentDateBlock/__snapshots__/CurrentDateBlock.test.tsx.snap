// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CURRENT-DATE-BLOCK TESTING Render current date 1`] = `
<div
  class="wrapper"
  data-testid="date-block"
>
  <button
    class="btn prevDateBtn gray"
    data-testid="prev-btn"
  />
  <div
    class="date gray"
    data-testid="date-elem"
  >
    <div
      class="calendarIcon"
    />
    <div
      class="dateText"
    >
      1 января
    </div>
  </div>
  <button
    class="btn nextDateBtn gray"
    data-testid="next-btn"
  />
</div>
`;
